src/main/java/com/mraof/minestuck/blockentity/IntellibeamLaserstationBlockEntity.java

```diff
@@ -0,0 +1,198 @@
+	}
+	
+	public void onRightClick(Player player)
+	{
+		ItemStack heldItem = player.getMainHandItem();
+		ItemStack taggedItem = AlchemyHelper.getDecodedItem(heldItem);
+		
+		if(isDecoded(taggedItem))
+		{
+			return;
+		}
+		if(card.isEmpty() && !this.hasCard() && !isDecoded(taggedItem))
+		{
+			tryInsertCard(heldItem);
+			return;
+		}
+		if(EXPERIENCE_LEVEL >= EXP_LEVEL_CAPACITY)
+		{
+			insertCard(decodeItem(taggedItem));
+			takeCard(player);
+			return;
+		}
+		addExperience(player);
+	}
+	
+	public void takeCard(Player player)
+	{
+		if (player.getMainHandItem().isEmpty())
+			player.setItemInHand(InteractionHand.MAIN_HAND, card);
+		else if (!player.getInventory().add(card))
+			dropCard(false, level, worldPosition, card);
+		else player.inventoryMenu.broadcastChanges();
+		
+		this.level.playSound(null, this.worldPosition, MSSoundEvents.INTELLIBEAM_LAZERSTATION_REMOVE_CARD.get(), SoundSource.BLOCKS, 0.5F, 1F);
+		
+		insertCard(ItemStack.EMPTY);
+	}
+	
+	public void tryInsertCard(ItemStack heldStack)
+	{
+		if (!heldStack.isEmpty() && heldStack.getItem() == MSItems.CAPTCHA_CARD.get())
+		{
+			insertCard(heldStack.split(1));
+			ItemStack cardStack = getCard();
+			ItemStack item = new ItemStack(MSBlocks.GENERIC_OBJECT.get());
+			
+			if (cardStack.hasTag() && cardStack.getTag().contains("contentID"))
+				item = AlchemyHelper.getDecodedItem(cardStack);
+		}
+	}
+	
+	public void dropCard(boolean inBlock, Level level, BlockPos pos, ItemStack item)
+	{
+		BlockPos dropPos;
+		if(inBlock)
+			dropPos = pos;
+		else if(!Block.canSupportCenter(level, pos.above(), Direction.DOWN))
+			dropPos = pos.above();
+		else dropPos = pos;
+		
+		Containers.dropItemStack(level, dropPos.getX(), dropPos.getY(), dropPos.getZ(), item);
+		
+	}
+	
+	public boolean hasCard()
+	{
+		return !this.getCard().isEmpty();
+	}
+	
+	public void insertCard(ItemStack card)
+	{
+		if (card.getItem() == MSItems.CAPTCHA_CARD.get() || card.isEmpty())
+		{
+			this.card = card;
+			if(level != null)
+			{
+				updateState();
+			}
+		}
+	}
+	
+	public ItemStack decodeItem(ItemStack taggedCard)
+	{
+		ItemStack stack = createEncodedItem(taggedCard, true);
+		stack.getOrCreateTag().putBoolean("decoded", true);
+		
+		EXPERIENCE_LEVEL = 0;
+		SOUND_SCALER = 0F;
+		
+		return stack;
+	}
+	
+	public void addExperience(Player player)
+	{
+		if (player.experienceLevel <= 0)
+		{
+			return;
+		}
+		if (player.getMainHandItem().isEmpty() && this.hasCard())
+		{
+			player.giveExperienceLevels(-1);
+			EXPERIENCE_LEVEL += 1;
+			SOUND_SCALER += 0.1;
+			
+			this.level.playSound(null, this.worldPosition, MSSoundEvents.INTELLIBEAM_LAZERSTATION_EXP_GATHER.get(), SoundSource.BLOCKS, 0.5F, 1F + SOUND_SCALER);
+		}
+	}
+	
+	public ItemStack getCard()
+	{
+		return this.card;
+	}
+	
+	public static boolean isDecoded(ItemStack decodedItem)
+	{
+		return decodedItem.getItem() == MSItems.CAPTCHA_CARD.get() && decodedItem.hasTag() && decodedItem.getTag().getBoolean("decoded");
+	};
```

Comment:
Looks a lot like `AlchemyHelper.isReadableCard()`
Perhaps you should use that one instead?

Reply:
done